sui_storage/src/lock_service.rs

```diff
@@ -0,0 +1,491 @@
+                    resp.send(self.delete_locks(&refs))
+                        .expect("Could not respond to sender!");
+                }
+            }
+        }
+        info!("LockService command loop stopped, the sender on other end hung up/dropped");
+    }
+
+    /// Loop to continuously process queries in a single thread
+    fn queries_loop(&self, receiver: Receiver<LockServiceQueries>) {
+        info!("LockService queries processing loop started");
+        while let Ok(msg) = receiver.recv() {
+            match msg {
+                LockServiceQueries::GetLock { object, resp } => {
+                    resp.send(self.get_lock(object))
+                        .expect("Could not respond to sender!");
+                }
+            }
+        }
+        info!("LockService queries loop stopped, the sender on other end hung up/dropped");
+    }
+}
+
+const LOCKSERVICE_QUEUE_LEN: usize = 500;
+
+/// Atomic Sui Object locking service.
+/// Primary abstraction is an atomic op to acquire a lock on a given set of objects.
+/// Atomicity relies on single threaded loop and only one instance per authority.
+#[derive(Clone)]
+pub struct LockService {
+    sender: SyncSender<LockServiceCommands>,
+    query_sender: SyncSender<LockServiceQueries>,
+}
+
+impl LockService {
+    /// Create a new instance of LockService.  For now, the caller has to guarantee only one per data store -
+    /// namely each SuiDataStore creates its own LockService.
+    pub fn new<P: AsRef<Path>>(path: P, db_options: Option<Options>) -> Result<Self, SuiError> {
+        let inner_service = LockServiceImpl::try_open_db(path, db_options)?;
+
+        // Now, create a sync channel and spawn a thread
+        let (sender, receiver) = sync_channel(LOCKSERVICE_QUEUE_LEN);
+        let inner2 = inner_service.clone();
+        std::thread::spawn(move || {
+            inner2.command_loop(receiver);
+        });
+
+        let (q_sender, q_receiver) = sync_channel(LOCKSERVICE_QUEUE_LEN);
+        std::thread::spawn(move || {
+            inner_service.queries_loop(q_receiver);
+        });
+
+        Ok(Self {
+            sender,
+            query_sender: q_sender,
+        })
+    }
+
+    /// Acquires a lock for a transaction on the given objects if they have all been initialized previously
+    /// to None state.  It is also OK if they have been set to the same transaction.
+    /// The locks are all set to the given transacton digest.
+    /// Otherwise, SuiError(TransactionLockDoesNotExist, ConflictingTransaction) is returned.
+    /// Note that this method sends a message to inner LockService implementation and waits for a response
+    pub async fn acquire_locks(
+        &self,
+        refs: Vec<ObjectRef>,
+        tx_digest: TransactionDigest,
+    ) -> SuiResult {
+        let (os_sender, os_receiver) = oneshot::channel::<SuiResult>();
+        // NOTE: below is blocking, switch to Tokio channels which are async?
```

Comment:
This is a must: use tokio async channel with an await on the sender, and using a blocking receive on the OS thread side: https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.blocking_recv

Reply:
@gdanezis thanks for the hint to use blocking_recv.  You might be interested to know that I at first tried making the loop async and using `recv().await` on every message receive, but that decreases the throughput quite a bit (maybe 7% overall on TPS).  So, a blocking receive single threaded loop is faster than a generic Tokio await loop.  Maybe the Tokio async await loop can be almost as fast with special executors....